src/geo/projection/projection.js

```diff
@@ -98,6 +98,14 @@ export default class Projection {
         }
     }
 
+    isPointAboveHorizon(tr: Transform, p: Point): boolean {
+        if (tr.elevation) {
+            return !this.pointCoordinate3D(tr, p.x, p.y);
+        }
```

Comment:
Is this valid for other projections such as mercator? Does `false` returned from `pointCoordinate3D` imply that the point is above horizon?

Reply:
> Is this valid for other projections such as mercator?

@mpulkki-mapbox Yes, this is valid for all non-globe projections. (I've moved this code from `transform.js`.)

> Does `false` returned from `pointCoordinate3D` imply that the point is above horizon?

Yes, it returns `null` in this case, passed from [`terrain.pointCoordinate`](https://github.com/mapbox/mapbox-gl-js/blob/51d405880c34606fb28b0cb411be03335f37d671/src/terrain/terrain.js#L1185).

Reply:
>>Is this valid for other projections such as mercator?
>
>@mpulkki-mapbox Yes, this is valid for all non-globe projections. (I've moved this code from transform.js.)

Yes I think this only gets resolved when terrain is enabled, refer [here](https://github.com/mapbox/mapbox-gl-js/blob/main/src/terrain/terrain.js#L1207). Otherwise we're always returning a valid coordinate, it's a difficult code path to follow because we're always checking against terrain. That's our full code path for non-globe projection:

```js
pointCoordinate3D(tr: Transform, x: number, y: number): ?Vec3 {
    const p = new Point(x, y);
    if (tr.elevation) {
        return tr.elevation.pointCoordinate(p); // null when dem tree raycast fails
    } else {
        const mc = this.pointCoordinate(tr, p.x, p.y, 0);
        return [mc.x, mc.y, mc.z]; // never null
    }
}
```